Projects/BookJSON2/main.go

```diff
@@ -30,26 +31,39 @@ var books []Book
 func (b *Books) book(w http.ResponseWriter, r *http.Request) {
 	// if r.Method == POST
 	switch r.Method {
-	case "POST": //Retriving a book with Id number
+	case "GET": //Retriving a book with Id number
 		{
+
 			w.Header().Set("content-type", "application/json")
-			jsonreqestfromweb := &Book{}
-			err := json.NewDecoder(r.Body).Decode(&jsonreqestfromweb)
+			val := r.FormValue("id")
+			id, err := strconv.Atoi(val)
 			if err != nil {
-				fmt.Println(err)
-				return
+				panic(err)
+			}
+			if id >= len(b.Books) {
+				w.WriteHeader(http.StatusNotFound)
+
+			} else {
+				json.NewEncoder(w).Encode(b.Books[id])
 			}
 
-			json.NewEncoder(w).Encode(b.Books[jsonreqestfromweb.Id])
-			break
 		}
 	case "PUT": // adding new book
 		{
 
 			w.Header().Set("content-type", "application/json")
 			newbook := Book{}
 			_ = json.NewDecoder(r.Body).Decode(&newbook)
-			b.Books = append(b.Books, newbook)
+			for _, item := range b.Books {
+				if newbook.Id == item.Id {
+					w.WriteHeader(http.StatusBadRequest)
+					break
+				} else {
+					b.Books = append(b.Books, newbook)
+					break
```

Comment:
This logic might be wrong.

If the books collection has 100 items, the loop will only iterate the first one. If the Id for the new book matches the first one in the array it will exit the loop. But if it does not match it, it will write it to the array and won't iterate the remaining 99 to check for duplicates.

Reply:
resolved with commit https://github.com/AmanSarraf/Go-Project/pull/9/commits/85fdee5b41960dddf4790f50c6def83cb4fe3ea5